### PR TITLE
Proposals update. Iteration 2. New staking implementation.

### PR DESCRIPTION
**Pull request by shamil-gadelshin**
This PR contains staking handler implementation for the proposals system. It implements staking via locks from the Substrate `balances` pallet.

Comments: 
- deleted joystream `stake` dependency from the proposals engine
- introduced `StakingManager` as an implementation of the `StakingHandler`
- member-to-staking-account-id is skipped for now

### CODE REVIEW

File: runtime-modules/proposals/engine/src/types/mod.rs [hunk 1]
```diff
@@ -419,13 +379,64 @@ pub struct ProposalCreationParameters<BlockNumber, Balance, MemberId, AccountId>
...
 
+/// Contains staking balance and account id.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, Default, Clone, Copy, PartialEq, Eq, Debug)]
+pub struct Stake<AccountId, Balance> {
+    /// Staking account id.
+    pub account_id: AccountId,
+
+    /// Staking balance.
+    pub balance: Balance,
```
Comment (bedeho) [inline runtime-modules/proposals/engine/src/types/mod.rs]: If staking for proposals is exact, then why do we need this?

File: runtime-modules/proposals/engine/src/lib.rs [hunk 2]
```diff
@@ -898,16 +845,101 @@ type ApprovedProposal<T> = ApprovedProposalData<
...
+    BalanceOf<T>,
     <T as system::Trait>::AccountId,
 >;
+
+pub struct StakingManager<T: Trait, LockId: Get<LockIdentifier>> {
+    trait_marker: PhantomData<T>,
+    lock_id_marker: PhantomData<LockId>,
+}
+
+impl<T: Trait, LockId: Get<LockIdentifier>> StakingHandler<T> for StakingManager<T, LockId> {
```
Comment (bedeho) [inline runtime-modules/proposals/engine/src/lib.rs:868]: This should not be in the engine itself, that makes the engine as a software module, tightly coupled to the balances and membership modules. It should be in the runtime, the same applies to the working group case.

Comment (bedeho) [inline runtime-modules/proposals/engine/src/lib.rs:868]: Why do we need to implement this again and again for each module, that seems wrong.

**Review by bedeho — CHANGES_REQUESTED** [review]

File: runtime-modules/proposals/engine/src/lib.rs (hunk 2 above)
Comment (shamil-gadelshin) [reply runtime-modules/proposals/engine/src/lib.rs:868]: I agree with both. It is situated here until we integrate at least 2 modules.

Comment (shamil-gadelshin) [reply runtime-modules/proposals/engine/src/lib.rs:868]: It is situated here until we integrate at least 2 modules. I don't expect copies with this handler.

File: runtime-modules/proposals/engine/src/types/mod.rs (hunk 1 above)
Comment (shamil-gadelshin) [reply runtime-modules/proposals/engine/src/types/mod.rs]: It is an artifact. We could keep it as a possible agility feature in order to make a stake 'equal or greater than' preset someday or remove it for all proposals.

File: runtime-modules/proposals/engine/src/lib.rs (hunk 2 above)
Comment (bedeho) [reply runtime-modules/proposals/engine/src/lib.rs:865]: OK, merging in that case.

File: runtime-modules/proposals/engine/src/types/mod.rs (hunk 1 above)
Comment (bedeho) [reply runtime-modules/proposals/engine/src/types/mod.rs]: No I don't think that is advisable, going to have to insist here.

File: runtime-modules/proposals/engine/src/lib.rs (hunk 2 above)
Comment (bedeho) [reply runtime-modules/proposals/engine/src/lib.rs:868]: Ok.

Comment (bedeho) [reply runtime-modules/proposals/engine/src/lib.rs:868]: Ok.

File: runtime-modules/proposals/engine/src/types/mod.rs (hunk 1 above)
Comment (shamil-gadelshin) [reply runtime-modules/proposals/engine/src/types/mod.rs]: Sure.

File: runtime-modules/proposals/engine/src/types/mod.rs [hunk 3]
```diff
@@ -379,15 +379,12 @@ pub struct ProposalCreationParameters<BlockNumber, Balance, MemberId, AccountId>
     pub exact_execution_block: Option<BlockNumber>,
 }
 
-/// Contains staking balance and account id.
+/// Contains staking account id.
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
 #[derive(Encode, Decode, Default, Clone, Copy, PartialEq, Eq, Debug)]
-pub struct Stake<AccountId, Balance> {
+pub struct Stake<AccountId> {
```
Comment (bedeho) [inline runtime-modules/proposals/engine/src/types/mod.rs]: :D Does this struct really make sense at this point! you are desperate not to go back to just a simple account? :P

**Review by bedeho — CHANGES_REQUESTED** [review]

File: runtime-modules/proposals/engine/src/types/mod.rs (hunk 3 above)
Comment (shamil-gadelshin) [reply runtime-modules/proposals/engine/src/types/mod.rs]: Ok. I removed it.

**Review by bedeho — APPROVED** [review]

**Pull request merged**